app/solvers/exact_solver.py

```diff
@@ -105,63 +59,106 @@ def duplicate(items):
 
 
-def return_the_best_value(dict):
-    naj = max(dict.keys())
+# returns the best profit from dictionary which consist of paths with profits
+def return_the_best_value(dict_temp):
+    naj = max(dict_temp.keys())
     return naj
 
 
-def return_cost_of_the_best_path(dict, best):
+# return the cost on the best (from profit side) path
+def return_cost_of_the_best_path(graph, dict_temp_two, best):
     w = 1
-    koszt = 0
-    do = len(dict[best]) - 1
-    for x in dict[best][:do]:
-        koszt = koszt + dict[x][dict[best][w]]
+    cost = 0
+    do = len(dict_temp_two[best]) - 1
+    for x in dict_temp_two[best][:do]:
+        cost = cost + graph[x][dict_temp_two[best][w]]
         w = w + 1
-    return koszt
+    return cost
 
-
-def create_answer_for_path_creation(dict, best):
+# create special form of answer for path creation
+def create_answer_for_path_creation(dict_temp_three, best):
     answer_plot = []
-    do = len(dict[best]) -1
+    do = len(dict_temp_three[best]) -1
     for x in range(do):
-        answer_plot.append((dict[best][x], dict[best][x + 1]))
+        answer_plot.append((dict_temp_three[best][x], dict_temp_three[best][x + 1]))
     return answer_plot
+
+
+def solve(cities: pd.DataFrame, paths: pd.DataFrame, time: pd.DataFrame):
+    assert isinstance(cities, pd.DataFrame), 'Wrong data format!'
+    assert isinstance(paths, pd.DataFrame), 'Wrong data format!'
+    assert isinstance(time, pd.DataFrame), 'Wrong data format!'
+
+    # finds all paths from chosen start point with eye on possible time
+    def find_all_possible_paths(graph, start, time_left, path=[], price=0):
+        nonlocal all_paths
+        time_left = time_left - price
+        path = path + [start]
+        if start not in graph:
+            return []
+        for node in graph[start].keys():
+            if time_left - graph[start][node] >= 0:
+                find_all_possible_paths(graph, node, time_left, path, graph[start][node])
+            elif path not in all_paths:
+                all_paths.append(path)
+            if time_left == 0:
+                break
+        return all_paths
+
+    # general paths creating for all possible starting points
+    def create_all_possible_paths(graph, time_at_the_beggining):
+        list_of_paths = []
+        for x in graph.keys():
+            # all_paths = []
+            list_of_paths = list_of_paths + find_all_possible_paths(graph, x, time_at_the_beggining)
+            print(x)
+        return list_of_paths
+
+    working_time = time['time'].values[0]
+
+    d = convert_to_dict(cities, paths)
+
+    all_paths = []
+    possible_paths = create_all_possible_paths(d, working_time)
+
+    cities_dict = create_cities_dictionary(d, cities)
+
+    cities_dict_with_values = choose_the_best_path(possible_paths, cities_dict)
+
+    best_profit = return_the_best_value(cities_dict_with_values)
+
+    cost_of_the_best_path = return_cost_of_the_best_path(d, cities_dict_with_values, best_profit)
```

Comment:
`cost_of_the_best_path` oznacza czas przejścia ścieżki?

Reply:
Brawo - starałem się żeby nazwy były choć trochę intuicyjne. Tak wylicza "faktyczny koszt" przejścia przez wybrane miasta

Reply:
Wciąż są kwiatki typu: `time_left = time_left - price`, czas minus pieniądze..?

Reply:
Tak i tutaj musi to zostać - w sensie ma być przekazywane przez argument w taki sposób. Price jako cena a nie pieniądze.

Reply:
Taka musi być nazwa zmiennej? `price`? Ten kod jest nieczytelny.